daemon/analytics.c

```diff
@@ -0,0 +1,195 @@
+#include "common.h"
+
+static const char *verify_required_directory(const char *dir) {
+    if(chdir(dir) == -1)
+        fatal("Cannot cd to directory '%s'", dir);
+
+    DIR *d = opendir(dir);
+    if(!d)
+        fatal("Cannot examine the contents of directory '%s'", dir);
+    closedir(d);
+
+    return dir;
+}
+
+static void get_system_timezone(void) {
+    // avoid flood calls to stat(/etc/localtime)
+    // http://stackoverflow.com/questions/4554271/how-to-avoid-excessive-stat-etc-localtime-calls-in-strftime-on-linux
+    const char *tz = getenv("TZ");
+    if(!tz || !*tz)
+        setenv("TZ", config_get(CONFIG_SECTION_GLOBAL, "TZ environment variable", ":/etc/localtime"), 0);
+
+    char buffer[FILENAME_MAX + 1] = "";
+    const char *timezone = NULL;
+    ssize_t ret;
+
+    // use the TZ variable
+    if(tz && *tz && *tz != ':') {
+        timezone = tz;
+        // info("TIMEZONE: using TZ variable '%s'", timezone);
+    }
+
+    // use the contents of /etc/timezone
+    if(!timezone && !read_file("/etc/timezone", buffer, FILENAME_MAX)) {
+        timezone = buffer;
+        // info("TIMEZONE: using the contents of /etc/timezone: '%s'", timezone);
+    }
+
+    // read the link /etc/localtime
+    if(!timezone) {
+        ret = readlink("/etc/localtime", buffer, FILENAME_MAX);
+
+        if(ret > 0) {
+            buffer[ret] = '\0';
+
+            char   *cmp    = "/usr/share/zoneinfo/";
+            size_t cmp_len = strlen(cmp);
+
+            char *s = strstr(buffer, cmp);
+            if (s && s[cmp_len]) {
+                timezone = &s[cmp_len];
+                // info("TIMEZONE: using the link of /etc/localtime: '%s'", timezone);
+            }
+        }
+        else
+            buffer[0] = '\0';
+    }
+
+    // find the timezone from strftime()
+    if(!timezone) {
+        time_t t;
+        struct tm *tmp, tmbuf;
+
+        t = now_realtime_sec();
+        tmp = localtime_r(&t, &tmbuf);
+
+        if (tmp != NULL) {
+            if(strftime(buffer, FILENAME_MAX, "%Z", tmp) == 0)
+                buffer[0] = '\0';
+            else {
+                buffer[FILENAME_MAX] = '\0';
+                timezone = buffer;
+                // info("TIMEZONE: using strftime(): '%s'", timezone);
+            }
+        }
+    }
+
+    if(timezone && *timezone) {
+        // make sure it does not have illegal characters
+        // info("TIMEZONE: fixing '%s'", timezone);
+
+        size_t len = strlen(timezone);
+        char tmp[len + 1];
+        char *d = tmp;
+        *d = '\0';
+
+        while(*timezone) {
+            if(isalnum(*timezone) || *timezone == '_' || *timezone == '/')
+                *d++ = *timezone++;
+            else
+                timezone++;
+        }
+        *d = '\0';
+        strncpyz(buffer, tmp, len);
+        timezone = buffer;
+        // info("TIMEZONE: fixed as '%s'", timezone);
+    }
+
+    if(!timezone || !*timezone)
+        timezone = "unknown";
+
```

Comment:
This function is using different methods to define the timezone, please, add a debug message that will give us conditions to identify the road used to get the timezone.

Reply:
Thiago! We actually had a look at this function, for a different reason, and I'd like your input (or anyone else's!). Please have in mind that this function was already there.

It seems to follow this path to get the timezone: `TZ` variable -> `/etc/timezone` -> `/etc/localtime` (and `strftime`, etc). (In the end of course everything is overwritten if the user has set the `timezone` option in `netdata.conf`.

I'm not sure if it is ok, mainly in the sense of whether `/etc/localtime` should be picked before `/etc/timezone`. `/etc/timezone` is in essence a file that could contain almost garbage, and it wouldn't really affect the local machines date. *Most* applications do not care of what is inside `/etc/timezone`. (And in e.g. macos, I think the file is not there).

The code to also pick up `/etc/localtime` expects it to be a link (to an entry in `/usr/share/zoneinfo`) but it's also not always the case. It could be a real file copied from there, so it won't work either then.

Plus, setting the `timezone` to `unknown`, I'm not sure it's ok. It should perhaps be set to `UTC`, which is also kinda the default for no timezone set.

Reply:
It looks like `/etc/timezone` is inside the function to keep compatibility with all possible operate systems, on Slackware and on FreeBSD we do not have this file.

When I gave the suggestion to add a debug message, I had in mind that some of these methods do not apply to some systems and it would be good to know which method was used on my computer, because sometimes users can adjust some group of variables or files, and forget others. My concern is Netdata to show different timezone when compared with other servers running on the same computer.

About the default timezone, like you said, I remember that PostgreSQL also uses UTC as [default timezone](https://www.postgresql.org/docs/13/datatype-datetime.html). I do not have any objection to use UTC as default, and I think it is the best option.

Reply:
IMO, the correct handling here should be, in order:

1. Any `timezone` value in `/etc/netdata.conf`. This obviously overrides any othe rsettings, so we should check it _first_
2. Any value in the `TZ` environment variable. This will be the norm for Docker containers, as well as being the de-facto standard for overriding whatever is set at the system level.
3. Any value in `/etc/timezone`. Only some Linux systems have this file, but those that do _should_ have it in sync with what `/etc/localtime` has, and it0s easier to parse (it will point to a path in the tzdata database).
4. If `/etc/localtime` is a link, the path it points to. This will usually only be the case if `/etc/timezone` _is not present_ (systems that have `/etc/timezone` tend to copy the tzdata entry, and use `/etc/timezone` to know which one to copy when the database gets updated).
5. Use `date '+%Z %z'` to get the timezone abbreviation and offset. This will (almost) always work, including if none of the above configuration is set up and `/etc/localtime` is an actual file instead of a link, and will return UTC if no other configuration exists.
6. In the exceedingly unusual case that none of the above works, continue to return ‘unknown’ as a way to indicate that something is wrong.

I agree with @thiagoftsm that we should indicate which method was used to determine the timezone, it’s useful debugging information.

Reply:
Hi guys, thanks for your input.

The `/etc/netdata.conf` does take precedence (if it's there, it will overwrite whatever it was picked up from the whole function, even if it's in the last line). From @Ferroin 's answer, I guess it already then works as it is. We could just re-enable the commented `info` there to have a log of what method was used.

Reply:
Hello @MrZammler ,

Are you going to add a message indicating the method used to define the timezone? 

Best regards!

Reply:
Yes! I've re-enabled the info messages with https://github.com/netdata/netdata/pull/10887/commits/c694b0f35be0c02ac6ef1966fd7450c9f629d730

Thanks!!